lib/rspec/core/output_wrapper.rb

```diff
@@ -0,0 +1,26 @@
+module RSpec
+  module Core
+    # @private
+    class OutputWrapper
+      # @private
+      attr_writer :output
+
+      # @private
+      def initialize(output)
+        @output = output
+      end
+
+      # Redirect calls for IO interface methods
+      (IO.instance_methods - Enumerable.instance_methods - Object.instance_methods + [:to_str]).each do |method|
+        define_method(method) do |*args, &block|
+          @output.send(method, *args, &block)
+        end
+      end
+
+      # @private
+      def respond_to?(*args)
+        args[0] == :output= || @output.respond_to?(*args)
+      end
```

Comment:
I agree, the wrapping should be called after a string became an `IO` object.
`output_stream` may be a string when it sent as a command line param:
Rspec --format progress --out specs.out

A description of  `Configuration#add_formatter#870` says:
      # @param output [String, IO] where the formatter will write its output.
      #   Can be an IO object or a string path to a file. ...

`Formatter::Loader#add` [contains](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb#L148) the code to check the `output_stream` interface:

```ruby
args = paths.map { |p| p.respond_to?(:puts) ? p : file_at(p) }
```

And [creates](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb#L256) file when the `output_stream` is a file path:

```ruby
def file_at(path)
  RSpec::Support::DirectoryMaker.mkdir_p(File.dirname(path))
  File.new(path, 'w')
end
```

I.e. `--out specs.out` creates `File.new('specs.out', 'w')`.

Now RSpec::Core::Formatters::Loader#add has two responsibilities:

1. To create a file if a `path` is a string
2. To register a formatter

We may move out creation of a file from `Loader#add`. For example on the level of `Configuration#add_formatter` when `Loader#add` always receives an `IO` object and `OutputWrapper` wraps only the `IO` interface.